drivers/SmartThings/sonos/src/init.lua

```diff
@@ -362,8 +374,81 @@ local function handle_ssdp_discovery(self, ssdp_group_info, callback)
   end
 end
 
+local ONE_HOUR_IN_SECONDS = 3600
+
+local oauth_token_tx, _ = cosock.channel.new()
+
 --- @type SonosDriver
 local driver = Driver("Sonos", {
+  waiting_for_token = false,
+  oauth_token_tx = oauth_token_tx,
+  oauth = {},
+  get_oauth_token_receive_handle = function(self)
+    oauth_token_tx:subscribe()
+  end,
+  get_oauth_token = function(self)
```

Comment:
`get_oauth_token` will only return a token if we actually *have* one. Otherwise, it'll trigger the OAuth activity in the hub firwmare, and immediately return with an error indicating why.